tutorials/tour/case-classes.md

```diff
@@ -10,80 +10,128 @@ tutorial-next: pattern-matching
-abstract class Term
-case class Var(name: String) extends Term
-case class Fun(arg: String, body: Term) extends Term
-case class App(f: Term, v: Term) extends Term
+abstract class Notification
+case class Email(sourceEmail: String, title: String, body: String) extends Notification
+case class SMS(sourceNumber: String, message: String) extends Notification
+case class VoiceRecording(contactName: String, link: String) extends Notification
 ```
 
-This class hierarchy can be used to represent terms of the [untyped lambda calculus](https://en.wikipedia.org/wiki/Lambda_calculus). To facilitate the construction of case class instances, Scala does not require that the `new` primitive is used. One can simply use the class name as a function.
+Instantiating a Case class is easy: (Note that we don't need to use the `new` keyword)
+
+```tut
+val emailFromJohn = Email("john.doe@mail.com", "Greetings From John!", "Hello World!")
+```
 
-Here is an example:
+The constructor parameters of case classes are treated as public values and can be accessed directly. 
 
 ```tut
-Fun("x", Fun("y", App(Var("x"), Var("y"))))
+val title = emailFromJohn.title
+println(title) // prints "Greetings From John!"
+```
+
+With case classes, you cannot mutate their fields directly. 
+
+```tut:fail
+emailFromJohn.title = "Goodbye From John!" // This is a compilation error. We cannot assign another value to val fields, which all case classes fields are by default.
 ```
 
-The constructor parameters of case classes are treated as public values and can be accessed directly.
+Instead, you make a copy using the `copy` method. As seen below, you can replace just some of the fields:
 
 ```tut
-val x = Var("x")
-println(x.name)
+val editedEmail = emailFromJohn.copy(title = "I am learning Scala!", body = "It's so cool!")
+
+println(emailFromJohn) // prints "Email(john.doe@mail.com,Greetings From John!,Hello World!)"
+println(editedEmail) // prints "Email(john.doe@mail.com,I am learning Scala,It's so cool!)"
 ```
 
 For every case class the Scala compiler generates an `equals` method which implements structural equality and a `toString` method. For instance:
 
 ```tut
-val x1 = Var("x")
-val x2 = Var("x")
-val y1 = Var("y")
-println("" + x1 + " == " + x2 + " => " + (x1 == x2))
-println("" + x1 + " == " + y1 + " => " + (x1 == y1))
+val firstSms = SMS("12345", "Hello!")
+val secondSms = SMS("12345", "Hello!")
+
+if (firstSms == secondSms) {
+  println("They are equal!")
+}
+
+println("SMS is: " + firstSms)
 ```
 
 will print
 
 ```
-Var(x) == Var(x) => true
-Var(x) == Var(y) => false
+They are equal!
+SMS is: SMS(12345, Hello!)
 ```
 
-It only makes sense to define case classes if pattern matching is used to decompose data structures. The following [object](singleton-objects.html) defines a pretty printer function for our lambda calculus representation:
+With Case Classes, all the necessary machinery required for using them in **Pattern Matching** is automatically generated by the compiler.
```

Comment:
Yeah I think I put that in because it was really useful to know that pattern matching isn't some magical thing Scala did which only applies to case classes (coming from a Java-ish background). However you are correct - if they read this tutorial on case classes before pattern matching (which is the right order, imo), then the implementation detail of pattern matching doesn't need to be mentioned.